contracts/IntervalTokenStore.sol

```diff
@@ -0,0 +1,142 @@
+    );
+
+    event Withdraw(
+        address user,
+        address token,
+        uint amount
+    );
+
+    // User => Token => BalanceState
+    mapping(address => mapping(address => BalanceState)) private balanceStates;
+
+    struct BalanceState {
+        uint256 balance;
+        PendingFlux pendingDeposits;
+        PendingFlux pendingWithdraws;
+    }
+
+    struct PendingFlux {
+        uint256 amount;
+        uint256 stateIndex;  // deposits will be processed for any currentStateIndex > stateIndex
+                              // withdraws are allowed for any currentStateIndex > stateIndex
+    }
+
+    uint256 public currentStateIndex=0;
+
+    function updateAndGetBalance(address user, address token) public returns(uint256) {
+        updateDepositsBalance(user, token);
+        uint balance = balanceStates[user][token].balance;
+        if (balanceStates[user][token].pendingWithdraws.stateIndex < currentStateIndex) {
+            balance -= Math.min(balanceStates[user][token].pendingWithdraws.amount, balance);
+        }
+        return balance;
+    }
+
+    function deposit(address token, uint amount) public {
+        updateDepositsBalance(msg.sender, token);
+        require(
+            ERC20(token).transferFrom(msg.sender, address(this), amount),
+            "Tokentransfer for deposit was not successful"
+        );
+        balanceStates[msg.sender][token].pendingDeposits.amount += amount;
+        balanceStates[msg.sender][token].pendingDeposits.stateIndex = currentStateIndex;
+        emit Deposit(msg.sender, token, amount, currentStateIndex);
+    }
+
+    function withdrawRequest(address token, uint amount) public {
+        balanceStates[msg.sender][token].pendingWithdraws = PendingFlux({ amount: amount, stateIndex: currentStateIndex });
+        emit WithdrawRequest(msg.sender, token, amount, currentStateIndex);
+    }
+
+    function withdraw(address token, uint amount) public {
+        updateDepositsBalance(msg.sender, token);
+
+        require(
+            balanceStates[msg.sender][token].pendingWithdraws.stateIndex < currentStateIndex,
+            "withdraw was not registered previously"
+        );
+
+        require(
+            balanceStates[msg.sender][token].pendingWithdraws.amount >= amount,
+            "registered withdraw-amount was not sufficient"
+        );
+
+        require(
+            balanceStates[msg.sender][token].balance >= amount,
+            "balances not sufficient"
+        );
+
+        balanceStates[msg.sender][token].balance -= amount;
+        delete balanceStates[msg.sender][token].pendingWithdraws;
+
+        ERC20(token).transfer(msg.sender, amount);
+        emit Withdraw(msg.sender, token, amount);
+    }
+
+    function updateDepositsBalance(address user, address token) public {
+        if (balanceStates[user][token].pendingDeposits.stateIndex < currentStateIndex) {
+            balanceStates[user][token].balance += balanceStates[user][token].pendingDeposits.amount;
```

Comment:
use SafeMath

Reply:
Technically, the safemaths are not needed here, but you are right, better safe than sorry later.